[lldb] Add flag to "settings show" to include default values

kastiglione opened the pull request:
Adds a `--defaults`/`-d` flag to `settings show`. This mode will _optionally_ show a setting's default value. In other words, this does not always print a default value for every setting.

A default value is not shown when the current value _is_ the default.

Note: some setting types do not print empty or invalid values. For these setting types, if the default value is empty or invalid, the same elision logic is applied to printing the default value.

kastiglione committed:
[lldb] Show setting default in new verbose mode

kastiglione committed:
paren init

kastiglione committed:
restore unintentionally deleted line

kastiglione committed:
rename string helper function

medismailben commented on lldb/source/Utility/ArchSpec.cpp:
```diff
@@ -1439,6 +1439,10 @@ bool lldb_private::operator==(const ArchSpec &lhs, const ArchSpec &rhs) {
   return lhs.GetCore() == rhs.GetCore();
 }
 
+bool lldb_private::operator!=(const ArchSpec &lhs, const ArchSpec &rhs) {
```
I thought this would be compiler generated if the opposite one was implemented. Is that not the case ?

kastiglione replied:
it seems not until c++20 https://en.cppreference.com/w/cpp/language/default_comparisons.html

kastiglione replied:
I'm not sure what the compiler is supposed to do, but without declaring and implementing `operator!=`, the compiler emits:

```
OptionValueArch.cpp:36:25: error: invalid operands to binary expression ('ArchSpec' and 'ArchSpec')
36 |         m_current_value != m_default_value && m_default_value.IsValid()) {
   |         ~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~
```

JDevlieghere commented on lldb/source/Interpreter/OptionValueArch.cpp:
```diff
@@ -30,6 +31,12 @@ void OptionValueArch::DumpValue(const ExecutionContext *exe_ctx, Stream &strm,
+
+    if (dump_mask & eDumpOptionDefaultValue &&
+        m_current_value != m_default_value && m_default_value.IsValid()) {
+      DefaultValueFormat label{strm};
```
It's mostly about the [Golden Rule](https://llvm.org/docs/CodingStandards.html#golden-rule): 

> If you are extending, enhancing, or bug fixing already implemented code, use the style that is already being used so that the source is uniform and easy to follow.

Cynically, one could interpret that as nothing can change, but hasn't been the case. We have made a bunch of code style changes when they offered sufficient value to "outweigh" the value brought by consistency. 

For braced initializes specifically, it turns out that we actually already have an entry in the style guide: https://llvm.org/docs/CodingStandards.html#do-not-use-braced-initializer-lists-to-call-a-constructor

DavidSpickett replied:
> I know some folks prefer braced initializers but it's not common in LLVM or LLDB except for initializer lists

Flang is a fanatic for braced init (https://flang.llvm.org/docs/C%2B%2Bstyle.html#c-language) but it is very much an outlier in llvm-project.

jimingham commented on lldb/source/Commands/CommandObjectSettings.cpp:
```diff
@@ -237,28 +237,62 @@ insert-before or insert-after.");
+                          ExecutionContext *execution_context) override {
+      const int short_option = m_getopt_table[option_idx].val;
+      switch (short_option) {
+      case 'v':
```
This is a tiny quibble, but unless we have other things in mind to include in a verbose listing, I'd suggest `--show-default` and `-d` rather than `--verbose` since that way we say exactly what we're doing.  I can't think of anything else we'd include here, so `--show-defaults` seems better to me.

jimingham replied:
And even if we do add extra stuff, seeing what the defaults are on their own seems like a useful enough action I expect people would want to dial it up on its own.

kastiglione replied:
what do you think of `--include-defaults` since "show" might imply that a default will be printed for every setting, but that's not what the PR currently does. Or are you also suggesting that this should be changed to show a default, even for cases where for example a setting type doesn't have a default value (array, dict, etc).

jimingham replied:
If we were to do separate displays for "defaults", "values" and "values  with defaults" then show would be appropriate.  But I think it's fine not to do that.  As you said below, you can run w/o an .lldbinit in command line to figure that out, and in Xcode you could do the same thing by making an empty target-specific lldb init file.  Presumably there's a way to do this in VSCode, but even if there weren't this is a special-purpose enough task having to do it with `-x` seems acceptable.

In that case, `--include-defaults` seems more accurate.

jimingham commented:
Do you think people are likely sometimes to want to see ONLY the default values?  Showing the settings can get pretty verbose, so if you were trying to see "how lldb looks w/o any user intervention" not having to see two values and figure out when having only one means the default is that, and sometimes it's the second one might make that easier.
If you wanted to do that, you'd have `--show-defaults` and `--show-values` and the former would default to false, and the latter to true.
But  I can't convince myself one way or the other whether that's really useful.  If this seems silly to you, feel free to ignore the question.

kastiglione commented:
> Do you think people are likely sometimes to want to see ONLY the default values?

@jimingham In that case I've used `lldb -x` (`--no-lldbinit`). Do you think that's good enough, or nah?

kastiglione committed:
Rename flag to --include-defaults

kastiglione commented on lldb/source/Commands/Options.td:
```diff
@@ -56,6 +56,11 @@ let Command = "settings clear" in {
 }
 
+let Command = "settings show" in {
+  def setshow_include_defaults : Option<"include-defaults", "d">,
```
Jim [originally suggested `--show-defaults`](https://github.com/llvm/llvm-project/pull/153233#discussion_r2271407473) too. I think "show" is shorter and cleaner, but I also think implies that the output will include a default value for every setting – however that's not what this change does. With that context, what are your thoughts?

kastiglione replied:
thought: since the subcommand is already `show`, does it make sense to have the flag be simply `--defaults`?

`show --show-defaults`
`show --defaults`

kastiglione replied:
changed to `settings show --defaults`.

labath reviewed (commented):
Makes sense to me. I'm wondering whether the two reasons for omitting the default value couldn't lead to confusion. Like, is the default value omitted because the current value *is* the default or because it's type does not support default values (which I guess means that the default values is empty/default constructed?).

However, that's just something to consider, I'm not insisting on a particular resolution to that.

kastiglione committed:
make DefaultValueFormat a class

kastiglione committed:
revert to original quoting style

jimingham commented:
I agree with Pavel that before is easier to read.

Another way to indicate that the aggregate values are empty is to print the empty version of however we allow you to specify elements, so in your case:

```
(lldb) settings show target.unset-env-vars
target.unset-env-vars (array of strings) = (default: [])
  [0]: "PATH"
  [1]: "PWD"
```

kastiglione committed:
simplify flag name to --defaults

kastiglione committed:
add tests for scalar settings types

kastiglione committed:
now add support for collection settings

kastiglione commented on lldb/include/lldb/Interpreter/OptionValue.h:
```diff
@@ -338,6 +340,14 @@ class OptionValue {
+      stream.PutCString(" (default: ");
+    }
+    ~DefaultValueFormat() { stream.PutChar(')'); }
+    Stream &stream;
```
done

kastiglione commented on lldb/source/Interpreter/OptionValueFormatEntity.cpp:
```diff
@@ -54,17 +55,18 @@ void OptionValueFormatEntity::DumpValue(const ExecutionContext *exe_ctx,
-    std::string escaped;
-    EscapeBackticks(m_current_format, escaped);
-    strm << '"' << escaped << '"';
+    strm.QuotedCString(EscapeBackticks(m_current_format).data());
```
reverted to the `<<` style

kastiglione commented:
Updated. Of note:
* collection types now show `(default: empty)`
* the flag is named `--defaults`, in part because `settings show --show-defaults` seemed redundant 
* added tests

slydiman commented:
The following buildbots are broken after this patch

[lldb-x86_64-win](https://lab.llvm.org/buildbot/#/builders/211/builds/1510)
[lldb-remote-linux-win](https://lab.llvm.org/buildbot/#/builders/197/builds/8267)

Please fix ASAP or revert to make them green.

kastiglione commented:
@slydiman I received emails yesterday and fixed the failures, but there's another failure that wasn't initially reported. I'll fix that now.

kastiglione commented:
@slydiman https://github.com/llvm/llvm-project/pull/154983
